kani-compiler/src/kani_middle/transform/check_uninit/ty_layout.rs

```diff
@@ -0,0 +1,328 @@
+            TyKind::RigidTy(rigid_ty) => match rigid_ty {
+                RigidTy::Str => {
+                    let slicee_ty = Ty::unsigned_ty(UintTy::U8);
+                    let size_in_bytes = slicee_ty.layout().unwrap().shape().size.bytes();
+                    let data_chunks = data_bytes_for_ty(&MachineInfo::target(), slicee_ty, 0)?;
+                    let layout = PointeeLayout::Slice {
+                        element_layout: generate_byte_mask(size_in_bytes, data_chunks),
+                    };
+                    Ok(PointeeInfo { pointee_ty: ty, layout })
+                }
+                RigidTy::Slice(slicee_ty) => {
+                    let size_in_bytes = slicee_ty.layout().unwrap().shape().size.bytes();
+                    let data_chunks = data_bytes_for_ty(&MachineInfo::target(), slicee_ty, 0)?;
+                    let layout = PointeeLayout::Slice {
+                        element_layout: generate_byte_mask(size_in_bytes, data_chunks),
+                    };
+                    Ok(PointeeInfo { pointee_ty: ty, layout })
+                }
+                RigidTy::Dynamic(..) => {
+                    Ok(PointeeInfo { pointee_ty: ty, layout: PointeeLayout::TraitObject })
+                }
+                _ => {
+                    if ty.layout().unwrap().shape().is_sized() {
+                        let size_in_bytes = ty.layout().unwrap().shape().size.bytes();
+                        let data_chunks = data_bytes_for_ty(&MachineInfo::target(), ty, 0)?;
+                        let layout = PointeeLayout::Sized {
+                            layout: generate_byte_mask(size_in_bytes, data_chunks),
+                        };
+                        Ok(PointeeInfo { pointee_ty: ty, layout })
+                    } else {
+                        Err(format!("Cannot determine type layout for type `{ty}`"))
+                    }
+                }
+            },
+            TyKind::Alias(..) | TyKind::Param(..) | TyKind::Bound(..) => {
+                unreachable!("Should only encounter monomorphized types at this point.")
+            }
+        }
+    }
+
+    pub fn ty(&self) -> &Ty {
+        &self.pointee_ty
+    }
+
+    pub fn layout(&self) -> &PointeeLayout {
+        &self.layout
+    }
+}
+
+/// Get a size of an initialized scalar.
+fn scalar_ty_size(
+    machine_info: &MachineInfo,
+    layout_shape: LayoutShape,
+    current_offset: usize,
+) -> DataBytes {
+    match layout_shape.abi {
+        ValueAbi::Scalar(Scalar::Initialized { value, .. }) => {
+            DataBytes { offset: current_offset, size: value.size(machine_info) }
+        }
+        ValueAbi::ScalarPair(
+            Scalar::Initialized { value: value_first, .. },
+            Scalar::Initialized { value: value_second, .. },
+        ) => DataBytes {
+            offset: current_offset,
+            size: MachineSize::from_bits(
+                value_first.size(machine_info).bits() + value_second.size(machine_info).bits(),
```

Comment:
What if these are not contiguous? Don't we need to generate two `DataBytes`?

Reply:
`scalar_ty_size` is only used for determining sizes of primitives and references / pointers. I believe `ValueAbi::ScalarPair` could not occur in the former case, and it is contiguous in the latter case.

That makes me think: should we just "inline" this function in two places and remove unnecessary match arms for each place?

Reply:
Inline makes sense. You could also add an assertion that the data is contiguous and add the your explanation as a comment.

Reply:
I inlined it into two places and actually retrieved the offsets in the pointer / ref case without assuming it is contiguous